### ISSUE TITLE
Minor frontend code changes result in huge diffs

### ISSUE THREAD

**Issue by znation** [apple/turicreate#1822]
When making even a very small frontend change, every single frontend asset seems to get regenerated and changes. As a result, the code diffs end up huge (and don't proportionately represent the amount that has changed). It's bad even that we're committing generated code to the repo at all; but changing all of it at this frequency is especially bad.

An example of a PR with only one line of frontend code change, but a huge diff that results:

https://github.com/apple/turicreate/pull/1818/files#diff-d5e25a42d06a8558328eab687e8fbb6bL213

**Comment by znation** [comment]
Work breakdown:

- [x] Integrate node.js into cmake build system under deps/src
- [x] Integrate libsass/sassc into cmake build system under deps/src
- [ ] Add source drops of all JS dependencies
- [x] Run TypeScript compiler on JS via CMake
- [ ] Run sassc compiler on (S)CSS via CMake

**Comment by znation** [comment]
WIP branch: https://github.com/znation/turicreate/tree/cmake_frontend_build

**Comment by TobyRoseman** [comment]
@znation - This is an important issue and I would like to get it resolved. All of the checkbox in this issue are done. However I believe there is still work to be done. Could you summarize the work that still needs to be completed?

**Comment by znation** [comment]
Remaining work:

- [ ] Fix CSS class name mangling, which breaks styling on almost all UI (revert the conversion to CSS Modules from CSS files and test).
- [ ] Test and fix issues in SFrame.explore UI
- [ ] Test and fix issues in annotation UI
- [ ] Test and fix issues in Evaluation.explore UI

**Comment by znation** [comment]
Unfortunately this turned out to be a huge amount of work (see the changes in the [in progress branch](https://github.com/znation/turicreate/tree/cmake_frontend_build)) with no end in sight -- those remaining work items are not trivial -- and I haven't had time to prioritize it.

**Comment by znation** [comment]
Another option for fixing this (without redoing the entire build system) is to depend on a recent version of Node.js/npm, and on internet access (!), at build time. We could fix this issue by taking on those dependencies without changing much in the way of code or configuration (this is what we currently do manually and commit the output to the repo).
